src/test/java/org/apache/sysml/test/integration/functions/binary/matrix/TransposeSelfMatrixMultiplication.java

```diff
@@ -0,0 +1,168 @@
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.sysml.test.integration.functions.binary.matrix;
+
+import java.util.HashMap;
+
+import org.junit.AfterClass;
+import org.junit.BeforeClass;
+import org.junit.Test;
+
+import org.apache.sysml.api.DMLScript;
+import org.apache.sysml.api.DMLScript.RUNTIME_PLATFORM;
+import org.apache.sysml.lops.LopProperties.ExecType;
+import org.apache.sysml.runtime.matrix.data.MatrixValue.CellIndex;
+import org.apache.sysml.test.integration.AutomatedTestBase;
+import org.apache.sysml.test.integration.TestConfiguration;
+import org.apache.sysml.test.utils.TestUtils;
+
+/**
+ * This test investigates the specific Hop-Lop rewrite t(X)%*%v -> t(t(v)%*%X).
+ * 
+ */
+public class TransposeSelfMatrixMultiplication extends AutomatedTestBase
+{
+
+        private final static String TEST_NAME1 = "TransposeSelfMatrixMultiplication";
+        private final static String TEST_DIR = "functions/binary/matrix/";
+        private final static String TEST_CLASS_DIR = TEST_DIR + TransposeSelfMatrixMultiplication.class.getSimpleName() + "/";
+        private final static double eps = 1e-10;
+
+        //multiblock
+        private final static int rowsA1 = 3;
+        private final static int colsA1 = 3;
+
+        //singleblock
+        private final static int rowsA2 = 2407;
+        private final static int colsA2 = 73;
+
+
+        private final static double sparsity1 = 0.7;
+        private final static double sparsity2 = 0.1;
+
+
+        @Override
+        public void setUp()
+        {
+                addTestConfiguration( TEST_NAME1,
+                        new TestConfiguration(TEST_CLASS_DIR, TEST_NAME1, new String[] { "C" }) );
+
+                if (TEST_CACHE_ENABLED) {
+                        setOutAndExpectedDeletionDisabled(true);
+                }
+        }
+
+        @BeforeClass
+        public static void init()
+        {
+                TestUtils.clearDirectory(TEST_DATA_DIR + TEST_CLASS_DIR);
+        }
+
+        @AfterClass
+        public static void cleanUp()
+        {
+                if (TEST_CACHE_ENABLED) {
+                        TestUtils.clearDirectory(TEST_DATA_DIR + TEST_CLASS_DIR);
+                }
+        }
+
+        @Test
+        public void testTransposeMMDenseDenseCP1()
+        {
+                runTransposeSelfMatrixMultiplication(false, ExecType.CP, false, true);
+        }
+
+        @Test
+        public void testTransposeMMDenseDenseCP2()
+        {
+                runTransposeSelfMatrixMultiplication(false, ExecType.CP, false, false);
+        }
+        
+        /**
+         * 
+         * @param sparseM1
+         * @param sparseM2
+         * @param instType
+         */
+        private void runTransposeSelfMatrixMultiplication( boolean sparseM1, ExecType instType, boolean vectorM2, boolean isLeftTransposed)
```

Comment:
Detailed comments here would be so helpful for other people to be able to add tests.